// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

edition = "2023";

package brain.evolution;

message RandomSearchAlgorithmSpec {
  int64 population_size = 1;  // Required.

  // Whether to remove the oldest or a random individual.
  bool remove_oldest = 2 [default = true];

  // If set, will sample individuals without replacement individuals when
  // indiciated by this spec. Otherwise, will sample with replacement.
  DistinctSelectionSpec distinct_selection = 3;
}

// Samples without replacement when the size of the population at or above
// `min_population_size`. Otherwise, samples with replacement.
message DistinctSelectionSpec {
  double min_population_size = 1;  // Required.
}
